File: bilibili-recommend-ui/app/src/main/java/com/bilibilirecommend/network/NetWork.java
package com.bilibilirecommend.network;

import android.util.Log;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.ResponseBody;
import retrofit2.CallAdapter;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * by y on 2016/9/13
 */
public class NetWork {

    private static final Converter.Factory gsonConverterFactory = GsonConverterFactory.create();
    private static final CallAdapter.Factory rxJavaCallAdapterFactory = RxJavaCallAdapterFactory.create();
    private static RecommendService apiService;

    private static OkHttpClient getOkHttp() {
        return new OkHttpClient.Builder()
                .retryOnConnectionFailure(true)
                .addInterceptor(new LogInterceptor())
                .build();
    }

    public static RecommendService getApi() {
        if (apiService == null) {
            apiService = getRetrofit(RecommendService.RECOMMEND_URL).create(RecommendService.class);
        }
        return apiService;
    }

    private static Retrofit getRetrofit(String url) {
        return new Retrofit.Builder()
                .client(getOkHttp())
                .baseUrl(url)
                .addConverterFactory(gsonConverterFactory)
                .addCallAdapterFactory(rxJavaCallAdapterFactory)
                .build();
    }

    private static class LogInterceptor implements Interceptor {
        @Override
        public okhttp3.Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            Log.i("LogUtils--> ", "request:" + request.toString());
            okhttp3.Response response = chain.proceed(chain.request());
            okhttp3.MediaType mediaType = response.body().contentType();
            String content = response.body().string();
            Log.i("LogUtils--> ", "response body:" + content);
            if (response.body() != null) {
                ResponseBody body = ResponseBody.create(mediaType, content);
                return response.newBuilder().body(body).build();
            } else {
                return response;
            }
        }
    }

}

File: BilibiliRecommendUi/app/src/main/java/com/bilibilirecommend/network/NetWork.kt
package com.bilibilirecommend.network

import android.util.Log
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException

object NetWork {

    private val gsonConverterFactory = GsonConverterFactory.create()
    private val rxJavaCallAdapterFactory = RxJavaCallAdapterFactory.create()
    private var apiService: RecommendService? = null

    private val okHttp: OkHttpClient
        get() = OkHttpClient.Builder()
                .retryOnConnectionFailure(true)
                .addInterceptor(LogInterceptor())
                .build()

    val api: RecommendService
        get() {
            if (apiService == null) {
                apiService = getRetrofit(RecommendService.RECOMMEND_URL).create(RecommendService::class.java)
            }
            return apiService!!
        }

    private fun getRetrofit(url: String): Retrofit {
        return Retrofit.Builder()
                .client(okHttp)
                .baseUrl(url)
                .addConverterFactory(gsonConverterFactory)
                .addCallAdapterFactory(rxJavaCallAdapterFactory)
                .build()
    }

    private class LogInterceptor : Interceptor {
        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): okhttp3.Response {
            val request = chain.request()
            Log.i("LogUtils--> ", "request:" + request.toString())
            val response = chain.proceed(chain.request())
            val mediaType = response.body().contentType()
            val content = response.body().string()
            Log.i("LogUtils--> ", "response body:$content")
            return if (response.body() != null) {
                val body = ResponseBody.create(mediaType, content)
                response.newBuilder().body(body).build()
            } else {
                response
            }
        }
    }

}
